### DISCUSSION TITLE
Coupling Variables Error

### DISCUSSION THREAD

**Discussion by FanZhangchen** [idaholab/moose#22103] [Q&A General] [answered]
Hello,
I am trying to implement a MOOSE framework with two governing equations and two nodal degrees of freedom. I try to use adCoupledVectorValue to represent the coupled variable. MOOSE could compile well, but when I tried to solve it, some errors came up.
`*** ERROR ***
The named variable is a standard variable, try a 'coupled[Value/Gradient/Dot/etc]...' function instead`
So I think it might be wrong with my coupling variables. I am new to MOOSE that I am not sure how to realize the problems with two coupling variables in two equations. Could you give me some suggestions about it? Thanks

**Comment by FanZhangchen** [comment]
The governing equations are attached here if you need them.
![image](https://user-images.githubusercontent.com/38935985/190406952-c135803f-7303-4827-9cbb-40c981a816f1.png)
![image](https://user-images.githubusercontent.com/38935985/190407008-1d274ebf-a6c0-4d75-940c-33917bf5feb5.png)
$\rho$ and $\rho_G$ are the nodal degrees of freedom.

**Comment by GiudGiud** [comment, marked as answer]
Hello

CoupledVectorValue is a routine for vector variables, not for regular single-component field variables.'
I think you want to use coupledValue or ADcoupledValue here
This retrieves the local quadrature point value of the variable passed

Guillaume

**Reply by FanZhangchen** [reply]
Hello

Thank you for your reply. I have tried to use ADcoupledValue and I got some errors. The error is attached here.
```
error: no match for 'operator-' (operand types are 'libMesh::CompareTypes<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true>, MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true> >::supertype' {aka 'MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true>'} and 'libMesh::boostcopy::enable_if_c<true, libMesh::TypeVector<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true> > >::type' {aka 'libMesh::TypeVector<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true> >'})
   27 |     - _grad_test[_i][_qp] * _rhot[_qp];
```
I think I might have some C++ syntax errors here. I will also attach my C++ file here. Do you have any ideas about my problem?

```
#include "ADGNDTrans.h"

registerMooseObject("framework2App", ADGNDTrans);

InputParameters
ADGNDTrans::validParams()
{
  auto params = ADKernel::validParams();
  params.addParam<Real>("backstress", 1.0, "Back stress' parameter");
  params.addClassDescription("The time derivative operator with the weak form of $(\\psi_i, "
                             "\\frac{\\partial rhog}{\\partial t})$.");
  params.addCoupledVar("rhot", "The variable representing the total dislocation density.");
  return params;
}

ADGNDTrans::ADGNDTrans(const InputParameters & parameters)
  : ADKernel(parameters),
    _backstress(getParam<Real>("backstress")),
    _rhot(adCoupledValue("rhot"))
{
}

ADReal
ADGNDTrans::computeQpResidual()
{
  return _backstress * _grad_u[_qp]*_grad_test[_i][_qp]
    - _grad_test[_i][_qp] * _rhot[_qp];
}
```

**Reply by GiudGiud** [reply]
yes the syntax is wrong. Basically the two terms being substracted are of a different type.
From the message, one is a single ADReal and the other is a vector of ADReals

**Reply by FanZhangchen** [reply]
Does it mean I need to add another kernel for these different types? Or the easy way to solve this error is to divide two terms into different kernels?

**Reply by GiudGiud** [reply]
no you dont, you need to check that the math makes sense here. You have a grad_u * grad_test and a grad_test * some_variable that is not a vector

I dont think that s right

**Reply by FanZhangchen** [reply]
Yes, I get your point. I will check if there is any math error. Thank you
